src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -129,33 +267,437 @@ public ResultSet getSchemas() throws SQLException {
+     * @return The return value of the Supplier.
+     * @throws SQLException If a RuntimeException that wraps a SQLException is caught.
+     */
+    private static <T> T liftSQLException(Supplier<T> f) throws SQLException {
+        try {
+            return f.get();
+        } catch (RuntimeException e) {
+            if (e.getCause() instanceof SQLException) {
+                throw (SQLException) e.getCause();
+            }
+            throw e;
+        }
+    }
+
+    // Helper class for representing all info needed to serialize column data for the
+    // getColumns and getColumnPrivileges methods. Intended for use with toGetColumnsDoc
+    // and toGetColumnPrivilegesDoc helpers.
+    private static class GetColumnsDocInfo {
+        String dbName;
+        String collName;
+        String columnName;
+        MongoJsonSchema columnSchema;
+        boolean isRequired;
+        int idx;
+
+        GetColumnsDocInfo(
+                String dbName,
+                String collName,
+                String columnName,
+                MongoJsonSchema columnSchema,
+                boolean isRequired,
+                int idx) {
+            this.dbName = dbName;
+            this.collName = collName;
+            this.columnName = columnName;
+            this.columnSchema = columnSchema;
+            this.isRequired = isRequired;
+            this.idx = idx;
+        }
+    }
+
+    // Helper for creating BSON documents for the getColumns method. Intended for use
+    // with the getColumnsFromDB helper method which is shared between getColumns and
+    // getColumnPrivileges.
+    private BsonDocument toGetColumnsDoc(GetColumnsDocInfo i) {
+        MongoSQLColumnInfo info;
+        try {
+            info = new MongoSQLColumnInfo(i.collName, i.columnName, i.columnSchema, i.isRequired);
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+
+        int nullability = info.getNullability();
+        String isNullable =
+                nullability == ResultSetMetaData.columnNoNulls
+                        ? "NO"
+                        : nullability == ResultSetMetaData.columnNullable ? "YES" : "";
+
+        return createBottomBson(
+                new BsonElement(TABLE_CAT, new BsonString(i.dbName)),
+                new BsonElement(TABLE_SCHEM, new BsonString("")),
+                new BsonElement(TABLE_NAME, new BsonString(i.collName)),
+                new BsonElement(COLUMN_NAME, new BsonString(i.columnName)),
+                new BsonElement(DATA_TYPE, new BsonInt32(info.getJDBCType())),
+                new BsonElement(TYPE_NAME, new BsonString(info.getTableAlias())),
+                new BsonElement(COLUMN_SIZE, new BsonNull()),
+                new BsonElement(BUFFER_LENGTH, new BsonInt32(0)),
+                new BsonElement(
+                        DECIMAL_DIGITS,
+                        new BsonNull()), // TODO DECIMAL_DIGITS -> Oliver's helper (may need to add new method)
```

Comment:
In #55 , Oliver is adding utility methods for getting this information. I think I'll need to add a couple more to get these last bits of information. That PR is very close to being ready so I'll just wait on that to finish up these TODOs.